internal/storage/image.go

```diff
@@ -398,6 +408,145 @@ func (svc *imageService) ImageStatus(systemContext *types.SystemContext, nameOrI
+
+	if err := svc.checkSignature(ctx, systemContext, policyContext, userSpecifiedImageParsed, localManifestDigest); err != nil {
+		return fmt.Errorf("checking signature of %q: %w", userSpecifiedImageParsed.String(), err)
+	}
+
+	return nil
+}
+
+func (svc *imageService) pullLocalManifestDigest(ctx context.Context, sys *types.SystemContext, policyContext *signature.PolicyContext, userSpecifiedImage reference.Named) (digest.Digest, error) {
+	srcRef, err := docker.NewReference(userSpecifiedImage)
+	if err != nil {
+		return "", fmt.Errorf("creating docker:// reference for %q: %w", userSpecifiedImage.String(), err)
+	}
+
+	destRef, err := istorage.Transport.NewStoreReference(svc.store, userSpecifiedImage, "")
+	if err != nil {
+		return "", fmt.Errorf("creating containers-storage: reference for %q: %w", userSpecifiedImage.String(), err)
+	}
+
+	manifestBytes, err := copy.Image(ctx, policyContext, destRef, srcRef, &copy.Options{
+		SourceCtx:      sys,
+		DestinationCtx: sys,
+		ReportWriter:   io.MultiWriter(os.Stdout), // The MultiWriter is a hack to hide os.Stdout being a TTY, so that we don't get progress bars and they don't interfere with the debug log
+	})
+	if err != nil {
+		var e errcode.Error
+		if as := errors.As(err, &e); as && e.ErrorCode().Descriptor().HTTPStatusCode == http.StatusUnauthorized {
+			// We cannot support private registries right now, but may with future support from the CRI.
+			log.Errorf(ctx, "Verifying policies against private registries it not supported yet: %v", err)
+			return "", nil
+		}
+
+		var policyErr signature.PolicyRequirementError
+		if errors.As(err, &policyErr) {
+			return "", fmt.Errorf("%v: %w", crierrors.ErrSignatureValidationFailed, err)
+		}
+
+		return "", fmt.Errorf("pulling %q to %q: %w", transports.ImageName(srcRef), transports.ImageName(destRef), err)
+	}
+	return digest.FromBytes(manifestBytes), nil
+}
+
+func (svc *imageService) checkSignature(ctx context.Context, sys *types.SystemContext, policyContext *signature.PolicyContext, userSpecifiedImage reference.Named, localManifestDigest digest.Digest) error {
+	userIdentityRef, err := docker.NewReference(userSpecifiedImage)
+	if err != nil {
+		return fmt.Errorf("creating docker:// reference for %q: %w", userSpecifiedImage.String(), err)
+	}
+
+	storeNamedReference, err := reference.WithDigest(reference.TrimNamed(userSpecifiedImage), localManifestDigest)
+	if err != nil {
+		return fmt.Errorf("creating a digested storage reference for %q + %q: %w", userSpecifiedImage, localManifestDigest, err)
+	}
+	storageRefWithoutID, err := istorage.Transport.NewStoreReference(svc.store, storeNamedReference, "")
+	if err != nil {
+		return fmt.Errorf("creating containers-storage: reference for %q: %w", storeNamedReference.String(), err)
+	}
+	log.Debugf(ctx, "Created storageRefWithoutID = %q", transports.ImageName(storageRefWithoutID))
+
+	// storageRefWithoutID should be good enough, but just to show the workflow with a known local image ID:
```

Comment:
Yeah, CRI-O should use the data it has available.

If it doesn’t have the local image ID, the extra lookup to get an ID and then look up by ID adds not value. OTOH the ImagePull “local image already present” shortcut might need to be ID-based to make sure the signature check, config digest check, and final return value all reliably refer to the same image.